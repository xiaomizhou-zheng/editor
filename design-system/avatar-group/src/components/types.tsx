import type { ElementType, ReactNode } from 'react';

import type { AnalyticsEvent } from '@atlaskit/analytics-next';
import type { AvatarPropTypes } from '@atlaskit/avatar';

import type { AvatarGroupItemProps } from './avatar-group-item';

export type DeepRequired<T> = {
  [P in keyof T]-?: Required<T[P]>;
};

export type AvatarProps = AvatarPropTypes & {
  name: string;
  // eslint-disable-next-line @repo/internal/react/boolean-prop-naming-convention
  enableTooltip?: boolean;
  key?: string | number;
};

export interface AvatarGroupOverrides {
  AvatarGroupItem?: {
    render?: (
      Component: ElementType<AvatarGroupItemProps>,
      props: AvatarGroupItemProps,
      index: number,
    ) => ReactNode;
  };
  Avatar?: {
    render?: (
      Component: ElementType<AvatarProps>,
      props: AvatarProps,
      index: number,
    ) => ReactNode;
  };
}

export type onAvatarClickHandler = (
  event: React.MouseEvent,
  analyticsEvent: AnalyticsEvent | undefined,
  index: number,
) => void;
