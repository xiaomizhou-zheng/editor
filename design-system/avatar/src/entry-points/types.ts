// eslint-disable-next-line @repo/internal/fs/filename-pattern-match
export type {
  AppearanceType,
  SizeType,
  IndicatorSizeType,
  AvatarClickEventHandler,
} from '../types';
