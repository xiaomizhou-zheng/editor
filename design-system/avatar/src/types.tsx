import { UIAnalyticsEvent } from '@atlaskit/analytics-next';

export type AvatarSizeMap = Record<SizeType, number>;
// eslint-disable-next-line @repo/internal/react/consistent-types-definitions
export type AppearanceType = 'circle' | 'square';

// eslint-disable-next-line @repo/internal/react/consistent-types-definitions
export type SizeType =
  | 'xsmall'
  | 'small'
  | 'medium'
  | 'large'
  | 'xlarge'
  | 'xxlarge';

// eslint-disable-next-line @repo/internal/react/consistent-types-definitions
export type IndicatorSizeType = 'small' | 'medium' | 'large' | 'xlarge';

export type AvatarClickEventHandler = (
  event: React.MouseEvent,
  analyticsEvent?: UIAnalyticsEvent,
) => void;
