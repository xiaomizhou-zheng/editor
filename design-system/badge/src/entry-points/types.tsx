export type { BadgeProps } from '../types';
