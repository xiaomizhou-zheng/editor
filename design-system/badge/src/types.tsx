import type { CSSProperties, ReactNode } from 'react';

export type ThemeAppearance =
  | 'added'
  | 'default'
  | 'important'
  | 'primary'
  | 'primaryInverted'
  | 'removed';

export interface BadgeProps {
  /**
   * Affects the visual style of the badge.
   */
  appearance?: ThemeAppearance;

  /**
   * The value displayed within the badge. A `ReactNode` can be provided for
   * custom-formatted numbers, however, badge should only be used in cases where you want to represent
   * a number.
   * Use a [lozenge](/packages/design-system/lozenge) for non-numeric information.
   */
  children?: number | ReactNode;

  /**
   * The maximum value to display. Defaults to `99`. If the value is 100, and max is 50, "50+" will be displayed.
   * This value should be greater than 0. If set to `false` the original value will be displayed regardless of
   * whether it is larger than the default maximum value.
   */
  max?: number | false;

  /**
   * Style customization to apply to the badge. Only `backgroundColor` and `color` are supported.
   */
  style?: Pick<CSSProperties, 'backgroundColor' | 'color'>;

  /**
   * A `testId` prop is provided for specified elements, which is a unique string that appears as a data attribute `data-testid` in the rendered code, serving as a hook for automated tests
   */
  testId?: string;
}
