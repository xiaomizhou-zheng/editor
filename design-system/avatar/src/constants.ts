// eslint-disable-next-line @repo/internal/fs/filename-pattern-match
import { gridSize } from '@atlaskit/theme/constants';

import { AvatarSizeMap, IndicatorSizeType } from './types';

export const AVATAR_SIZES: AvatarSizeMap = {
  xsmall: gridSize() * 2,
  small: gridSize() * 3,
  medium: gridSize() * 4,
  large: gridSize() * 5,
  xlarge: gridSize() * 12,
  xxlarge: gridSize() * 16,
};

// border radius only applies to "square" avatars
export const AVATAR_RADIUS: AvatarSizeMap = {
  xsmall: 2,
  small: 2,
  medium: 3,
  large: 3,
  xlarge: 6,
  xxlarge: 12,
};

export const BORDER_WIDTH = 2;
export const ACTIVE_SCALE_FACTOR = 0.9;

export const ICON_SIZES: Record<IndicatorSizeType, number> = {
  small: 12,
  medium: 14,
  large: 15,
  xlarge: 18,
};

export const ICON_OFFSET: Record<IndicatorSizeType, number> = {
  small: 0,
  medium: 0,
  large: 1,
  xlarge: 7,
};
