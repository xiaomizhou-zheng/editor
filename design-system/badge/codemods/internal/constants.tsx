export const BADGE_PACKAGE_NAME = '@atlaskit/badge';
export const APPEARANCE_PROP_NAME = 'appearance';
export const STYLE_PROP_NAME = 'style';
