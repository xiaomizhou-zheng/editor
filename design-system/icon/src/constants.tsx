import type { Size } from './types';

export const sizes: Record<Size, string> = {
  small: '16px',
  medium: '24px',
  large: '32px',
  xlarge: '48px',
};

export const sizeMap: Record<Size, Size> = {
  small: 'small',
  medium: 'medium',
  large: 'large',
  xlarge: 'xlarge',
};

export const dimensions = {
  small: {
    width: sizes.small,
    height: sizes.small,
  },
  medium: {
    width: sizes.medium,
    height: sizes.medium,
  },
  large: {
    width: sizes.large,
    height: sizes.large,
  },
  xlarge: {
    width: sizes.xlarge,
    height: sizes.xlarge,
  },
} as const;
