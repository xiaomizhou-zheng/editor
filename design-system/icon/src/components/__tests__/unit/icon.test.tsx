import React, { CSSProperties } from 'react';
import { cleanup, render } from '@testing-library/react';

import Icon, { size, CustomGlyphProps, IconProps } from '../../..';
import AddIcon from '../../../../glyph/add';
import { sizes as sizeValues } from '../../../constants';
import { Size } from '../../../types';

describe('@atlaskit/icon', () => {
  afterEach(cleanup);
  describe('Icon', () => {
    const secretContent = 'secret content';
    const secretWrapper = (props: CustomGlyphProps) => (
      // eslint-disable-next-line @repo/internal/react/no-unsafe-spread-props
      <svg {...props}>{secretContent}</svg>
    );
    // eslint-disable-next-line @repo/internal/react/no-unsafe-spread-props
    const empty = (props: CustomGlyphProps) => <svg {...props}>Icon</svg>;
    const MyIcon = (props: IconProps) => (
      // eslint-disable-next-line @repo/internal/react/no-unsafe-spread-props
      <Icon glyph={secretWrapper} {...props} />
    );

    it('should match the DOM Snapshot', () => {
      const { getByLabelText } = render(<Icon glyph={empty} label="My icon" />);

      expect(getByLabelText('My icon')).toMatchSnapshot();
    });

    describe('glyph prop', () => {
      it('should render an SVG provided via JSX', () => {
        const id = 'customSvg';
        const customGlyphJsx = (props: {}) => (
          // eslint-disable-next-line @repo/internal/react/no-unsafe-spread-props
          <svg {...props} data-testid={id} />
        );

        const { getByTestId } = render(
          <Icon glyph={customGlyphJsx} label="" />,
        );
        expect(getByTestId(id)).toBeDefined();
      });

      it('should present itself as an image', () => {
        const { getByRole } = render(
          <Icon glyph={empty} testId="empty-icon" label="My icon" />,
        );

        expect(getByRole('img')).toBeDefined();
      });
    });

    describe('dangerouslySetGlyph prop', () => {
      const id = 'customSvg';
      const customGlyphString = `<svg data-testid=${id}></svg>`;

      it('should render an SVG provided as a string', () => {
        const { getByTestId } = render(
          <Icon
            testId="test-icon"
            dangerouslySetGlyph={customGlyphString}
            label="hello-world"
          />,
        );
        const svg = getByTestId(id);
        expect(svg).toBeDefined();
        expect(svg.nodeName).toEqual('svg');
      });

      it('should present itself as an image when label is defined', () => {
        const testId = 'test-icon';
        const { getByTestId } = render(
          <Icon
            testId="test-icon"
            dangerouslySetGlyph={customGlyphString}
            label="hello-world"
          />,
        );

        const element = getByTestId(testId);
        expect(element.getAttribute('role')).toEqual('img');
        expect(element.getAttribute('aria-label')).toEqual('hello-world');
      });

      it('should present as hidden, without a role when the label is an empty string', () => {
        const testId = 'test-icon';
        const { getByTestId } = render(
          <Icon
            testId="test-icon"
            dangerouslySetGlyph={customGlyphString}
            label=""
          />,
        );

        const element = getByTestId(testId);
        expect(element).not.toHaveAttribute('role'); // the default role for a span is `none`
        expect(element).not.toHaveAttribute('aria-label');
        expect(element).toHaveAttribute('aria-hidden', 'true');
      });
    });

    describe('exports', () => {
      it('exports handled sizes', () => {
        expect(size).not.toBe(undefined);
        expect(
          (Object.keys(size) as (keyof typeof size)[]).map((key) => size[key]),
        ).toEqual(['small', 'medium', 'large', 'xlarge']);
      });
    });

    it('should be possible to create an Icon via a subclass', () => {
      const { getByLabelText } = render(<MyIcon label="My icon" />);

      expect(getByLabelText('My icon')).toBeDefined();
    });

    describe('size property', () => {
      const sizes: Size[] = ['small', 'medium', 'large', 'xlarge'];

      sizes.forEach((s) => {
        it(`with value ${s}`, () => {
          const { getByLabelText } = render(
            <Icon glyph={empty} label={s} size={s} />,
          );
          const element = getByLabelText(s);
          expect(element).toHaveStyleDeclaration('height', sizeValues[s]);
          expect(element).toHaveStyleDeclaration('width', sizeValues[s]);
        });
      });

      it(`should use width/height if provided`, () => {
        const label = 'width';
        const props = { width: 10, height: 10 } as CSSProperties;
        const { getByLabelText } = render(
          <Icon glyph={empty} label={label} {...props} />,
        );
        const element = getByLabelText(label);
        expect(element).toHaveStyleDeclaration('height', '10px');
        expect(element).toHaveStyleDeclaration('width', '10px');
      });

      it(`should use width/height above size`, () => {
        const size = 'large';
        const props = { width: 10, height: 10 } as CSSProperties;
        const { getByLabelText } = render(
          <Icon glyph={empty} label={size} size={size} {...props} />,
        );
        const element = getByLabelText(size);
        expect(element).toHaveStyleDeclaration('height', '10px');
        expect(element).toHaveStyleDeclaration('width', '10px');
      });
    });

    describe('primaryColor property', () => {
      const testLabel = 'test';
      it('is set to inherit the text color by default', () => {
        const { getByLabelText } = render(<MyIcon label={testLabel} />);

        expect(getByLabelText(testLabel).firstChild).toHaveStyle(
          `color: var(--icon-color)`,
        );
      });
    });

    describe('secondaryColor property', () => {
      it('is set to the default theme background color by default', () => {
        const label = 'default secondaryColor';
        const { getByLabelText } = render(<MyIcon label={label} />);

        expect(getByLabelText(label).firstChild).toHaveStyle(
          `fill: var(--icon-secondary-color)`,
        );
      });

      it('can be changed to a hex value', () => {
        const secondaryColor = '#ff0000';
        const label = 'hex secondaryColor';
        const { getByLabelText } = render(
          <MyIcon label={label} secondaryColor={secondaryColor} />,
        );

        expect(getByLabelText(label).firstChild).toHaveStyle(
          `fill: var(--icon-secondary-color)`,
        );
      });

      it('can be changed to a named color', () => {
        const secondaryColor = 'rebeccapurple';
        const label = 'hex secondaryColor';
        const { getByLabelText } = render(
          <MyIcon label={label} secondaryColor={secondaryColor} />,
        );

        const element = getByLabelText(label);

        expect(element.firstChild).toHaveStyle(
          'fill: var(--icon-secondary-color)',
        );
      });
    });
  });
  describe('Glyph', () => {
    afterEach(() => {
      jest.resetAllMocks();
    });

    const glyph = <AddIcon testId="test" label="test-label" />;

    it('should match snapshot', () => {
      const { getByTestId } = render(glyph);
      expect(getByTestId('test')).toMatchSnapshot();
    });

    it('should have the correct label', () => {
      const { getByLabelText } = render(glyph);
      expect(getByLabelText('test-label')).toBeDefined();
    });
  });
});
