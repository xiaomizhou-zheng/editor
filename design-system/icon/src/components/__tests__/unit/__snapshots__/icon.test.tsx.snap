// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@atlaskit/icon Glyph should match snapshot 1`] = `
.emotion-0 {
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  line-height: 1;
}

.emotion-0 >svg {
  overflow: hidden;
  pointer-events: none;
  max-width: 100%;
  max-height: 100%;
  color: var(--icon-primary-color);
  fill: var(--icon-secondary-color);
  vertical-align: bottom;
}

.emotion-0 >svg stop {
  stop-color: currentColor;
}

@media screen and (forced-colors: active) {
  .emotion-0 >svg {
    -webkit-filter: grayscale(1);
    filter: grayscale(1);
    --icon-primary-color: CanvasText;
    --icon-secondary-color: Canvas;
  }
}

<span
  aria-label="test-label"
  class="emotion-0"
  data-testid="test"
  role="img"
  style="--icon-primary-color: currentColor; --icon-secondary-color: var(--ds-surface, #FFFFFF);"
>
  <svg
    height="24"
    role="presentation"
    viewBox="0 0 24 24"
    width="24"
  >
    <path
      d="M13 11V3.993A.997.997 0 0012 3c-.556 0-1 .445-1 .993V11H3.993A.997.997 0 003 12c0 .557.445 1 .993 1H11v7.007c0 .548.448.993 1 .993.556 0 1-.445 1-.993V13h7.007A.997.997 0 0021 12c0-.556-.445-1-.993-1H13z"
      fill="currentColor"
      fill-rule="evenodd"
    />
  </svg>
</span>
`;

exports[`@atlaskit/icon Icon should match the DOM Snapshot 1`] = `
.emotion-0 {
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  line-height: 1;
}

.emotion-0 >svg {
  overflow: hidden;
  pointer-events: none;
  max-width: 100%;
  max-height: 100%;
  color: var(--icon-primary-color);
  fill: var(--icon-secondary-color);
  vertical-align: bottom;
}

.emotion-0 >svg stop {
  stop-color: currentColor;
}

@media screen and (forced-colors: active) {
  .emotion-0 >svg {
    -webkit-filter: grayscale(1);
    filter: grayscale(1);
    --icon-primary-color: CanvasText;
    --icon-secondary-color: Canvas;
  }
}

<span
  aria-label="My icon"
  class="emotion-0"
  role="img"
  style="--icon-primary-color: currentColor; --icon-secondary-color: var(--ds-surface, #FFFFFF);"
>
  <svg
    role="presentation"
  >
    Icon
  </svg>
</span>
`;
